src/certificates_helpers.py

```diff
@@ -0,0 +1,167 @@
+    """Generate a certificate and private key.
+
+    Args:
+        common_name (str): Common name for the certificate
+        sans_ips (list[str]): List of Subject Alternative Names (IPs)
+        ca_common_name (str): Common name for the CA certificate
+        validity (int): Certificate validity time (in days). The same value
+            is used for the CA certificate.
+
+    Returns:
+        Tuple[str, str, str]: Certificate, CA Certificate, Private Key
+    """
+    private_key = _generate_private_key()
+    ca_key = _generate_private_key()
+    ca_certificate = _generate_ca_certificate(
+        private_key=ca_key, common_name=ca_common_name, validity=validity
+    )
+    certificate = _generate_certificate(
+        private_key=private_key,
+        common_name=common_name,
+        sans_ips=sans_ips,
+        ca_cert=ca_certificate,
+        ca_key=ca_key,
+        validity=validity,
+    )
+    return certificate, ca_certificate, private_key
+
+
+def certificate_issuer_has_common_name(certificate: str, common_name: str) -> bool:
+    """Check if a certificate has a specific common name."""
+    try:
+        loaded_certificate = x509.load_pem_x509_certificate(certificate.encode())
+        issuer_common_name = loaded_certificate.issuer.get_attributes_for_oid(
+            x509.NameOID.COMMON_NAME
+        )[0]
+    except (ValueError, TypeError):
+        logger.warning("Failed to load certificate")
+        return False
+    return issuer_common_name.value == common_name
+
+
+def _generate_private_key() -> str:
+    private_key = rsa.generate_private_key(
+        public_exponent=PUBLIC_EXPONENT,
+        key_size=KEY_SIZE,
+    )
+    key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+    return key_bytes.decode().strip()
+
+
+def _generate_certificate(
+    private_key: str,
+    common_name: str,
+    sans_ips: List[str],
+    ca_cert: str,
+    ca_key: str,
+    validity: int,
+) -> str:
+    """Generate a self-signed certificate directly."""
+    private_key_obj = serialization.load_pem_private_key(private_key.encode(), password=None)
+    assert isinstance(private_key_obj, rsa.RSAPrivateKey)
+    ca_cert_obj = x509.load_pem_x509_certificate(ca_cert.encode())
+    subject = x509.Name([x509.NameAttribute(x509.NameOID.COMMON_NAME, common_name)])
+    cert_builder = (
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(ca_cert_obj.issuer)
+        .public_key(private_key_obj.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.now(timezone.utc))
+        .not_valid_after(datetime.now(timezone.utc) + timedelta(days=validity))
+        .add_extension(
+            x509.SubjectAlternativeName(
+                [x509.IPAddress(ipaddress.ip_address(san)) for san in sans_ips]
```

Comment:
Should we add the common name as DNS here too? That seems to be standard now.

Reply:
Done, added an argument to pass in the sans_dns